juju/juju
# Save and delete machine rows in dqlite as machines are saved to mongo

We need to have machine rows in dqlite so that block devices and other dependent entities which require a FK to the machine table can be written.

Here we do the minimal wiring of the domain machine service into the bits of the juju infrastructure to allow that double writing to occur. This won't be the end story - it's just the most minimally invasive approach to do the job while we transition. The key insertion points are:
- deploy application
- add unit
- add machine
- machine undertaker
- cleanup containers

As an experiment, a `WithMachineService` method is added to the test service factory so that it doesn't return nil all the time for the machine service in tests.

The schemagen machinery needed updating also, since the nil service factory has started to cause issues and needed to be supplied a test version.

## QA steps

bootstrap and run through the key scenarios for machine addition

- juju add-machine
- juju add-machine lxd
- juju deploy foo
- juju add-unit foo
- juju deploy bar --to N
- juju deploy bar --to lxd:N
- juju deploy bar --to lxd

Also juju-remove container(s) and machine(s)

For each case, select from the dqlite machine table and check that there's a row with each of the machine ids in there.

**Jira card:** JUJU-5219
